scala-thoughts.md

```diff
@@ -0,0 +1,9 @@
+
+In terms of the language itself, Scala to me seems to be a mix between Haskell and Java, which sounds really weird. However, as a result of this, the language that I most easily compare it to is actually python, which can do both the object oriented code style and functional style very easily. With this as the basis for comparison, it seems to me that Scala is more flexible than Java in terms of OOP programming, but less flexible than python in terms of functional programming.
+
+This isn't strictly true, so I'll expand on it a little bit. When I say that Scala is more flexible than Java, I mean that it lets you do a lot more with OOP, using built in factory methods, traits, and non-OOP scripts in the same place. This is something that I like in python a lot - mixing one-time scripts with classes in a large project. It seems though that Scala does this a lot better than Python. Since its object oriented, it can detect easy mistakes at compile time, and is better for making large projects. Since OOP in python is a little hacky, I like this aspect of Scala a lot more, and I'd rather write a large OOP project in Scala once I get used to it. However, I found with Scala for this assignment that the vast majority of the issues that I was having were with getting the code to compile due to type errors. Part of this I'm sure is the lack of familiarity with the language. At the same time, going back to the comparison with Python, I use functional programming constructs in Python so often that having to think about how to actually manage the types in the higher order functions when I already knew exactly how to code the solution from a logical standpoint was uncomfortable. Unfortunately, part of this seems to be just a construct of the type system in Scala that Python doesn't have. In Python, you can map over lists of arbitrary things without question, doing random things that shouldn't type check in any sane language, but that you know will work because of specific understanding of what's going on. While I'm sure you *can* do this in Scala, it would involve quite a bit of type weirdness. For this reason, I say that Scala is less flexible than Python for functional programming.
```

Comment:
Nice analysis! As someone who does mostly Java, I had a different experience, but I can definitely see where you're coming.